src/webgpu/shader/validation/expression/call/builtin/pow.spec.ts

```diff
@@ -0,0 +1,175 @@
+const builtin = 'pow';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  Type,
+  kConcreteIntegerScalarsAndVectors,
+  kConvertableToFloatScalarsAndVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatScalarsAndVectors);
+
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<number> {
+  switch (type) {
+    case Type.f32:
+      return quantizeToF32;
+    case Type.f16:
+      return quantizeToF16;
+    default:
+      return (v: number) => v;
+  }
+}
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() rejects invalid values
+
+TODO(http://github.com/gpuweb/issues/4527): This validation matches what is currently in Tint but
+it needs to be clarified in the spec if this is the desired behavior.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    let expectedResult = true;
+
+    const a = Number(t.params.a);
+    const b = Number(t.params.b);
+    if (a < 0 || (a === 0 && b <= 0)) {
```

Comment:
LGTM to land.

This matches GLSL exactly.

Also, MSL says for both pow and powr:
"Implemented as exp2(y * log2(x)).
Undefined for x = 0 and y = 0"
Because it uses log2(x) , then x < 0 is disallowed.

HLSL doesn't say.

IEEE 754-2018 defines pown, pow, and powr.  
*  pown: exponent is an integer
* pow  exponent y is real, but must agree with pown when y is integral.  Domain is [-inf,+inf]x[-inf,+inf]
* powr is implemented as exp( y * log(x)) so domain is [ 0, +inf ] x [-inf, +inf].
Closest we have in shader languages is powr.

Taking the intersection of GLSL, MSL, and IEEE powr, we end up using GLSL's rule.

So LGTM to land.